HW2/P4/P4.txt

```diff
@@ -0,0 +1,131 @@
+
+For the asynchronous filtering version:
+
+	Number of threads: 1
+	3.38918185234 seconds for 10 filter passes.
+	Number of threads: 2
+	1.59523296356 seconds for 10 filter passes.
+	Number of threads: 4
+	1.41955685616 seconds for 10 filter passes.
+	Number of threads: 8
+	1.40394091606 seconds for 10 filter passes.
+
+First, it is no surprise that the performance for the one-threaded case is
+essentially the same in both filtering versions, since there are no parallel
+threads executing.
+
+The performance for the two-threaded version is effectively the same for
+both synchronous and asynchronous versions, which is also expected since each
+thread must wait for the other thread to finish processing each iteration,
+as there are only two threads.
+
+It looks like we do see some savings, mostly in the four-threaded case, when 
+threads only have to wait for neighboring threads (as opposed to all active
+threads) to complete before continuing. 
+
+Note that while there is a marked improvement when moving from one to two
+threads in both cases, the improvement when moving from two to four
+threads is not substantial at all. As noted in class, the GIL limits
+improvements in performance (see also NewGIL.pdf reference), and we probably
+need to drop down to C-level code and constructs that allow us to run without
+the GIL before seeing more marked improvement. There is also overhead
+associated with thread communication, and this does not necessarily get more
+efficient with a greater number of threads.
+
+Moreover, given that all threads must wait for the others to complete an
+iteration, the speed of any given thread is bound by that of the slowest one
+(usually the one that has to do the most work) so there is potentially much
+idling.
+
+4. Additional experimentation
+
+We also experimented with simply creating new threads for each iteration 
+of the loop in the original code. Threads would complete their task, join to
+return their results, and presumably terminate, with new threads being 
+created for the next iteration. This potentially means we would be creating
+hundreds, perhaps thousands of threads, depending on the number of iterations
+and threads we are creating. 
+
+    for i in range(iterations):
+        threadlist = []
+        for threadidx in range(num_threads):
+            th = threading.Thread(target=filtering.median_3x3,
+                                  args=(tmpA, tmpB, threadidx, num_threads))
+            threadlist.append(th)
+            th.start()
+
+        for thread in threadlist:
+            thread.join()
+
+        tmpA, tmpB = tmpB, tmpA
+
+It was instructive to note that the performance without using any
+communication constructs (other than what is implicit in join) was actually 
+competitive with the synchronous case, particularly when four threads were
+used. This may provide further evidence that there is clearly overhead
+associated with communication costs between threads. 
```

Comment:
Good point to call out. I figured this was the case but didn't think to try to quantify it.